src/core/dense_set.h

```diff
@@ -0,0 +1,473 @@
+    }
+
+    bool IsObject() const {
+      return (uintptr_t(ptr_) & kLinkBit) == 0;
+    }
+
+    bool IsLink() const {
+      return (uintptr_t(ptr_) & kLinkBit) == kLinkBit;
+    }
+
+    bool IsEmpty() const {
+      return ptr_ == nullptr;
+    }
+
+    void* Raw() const {
+      return (void*)(uintptr_t(ptr_) & ~kTagMask);
+    }
+
+    bool IsDisplaced() const {
+      return (uintptr_t(ptr_) & kDisplaceBit) == kDisplaceBit;
+    }
+
+    void SetLink(DenseLinkKey* lk) {
+      ptr_ = (void*)(uintptr_t(lk) | kLinkBit);
+    }
+
+    void SetDisplaced(int direction) {
+      DCHECK(direction == -1 || direction == 1);
+      ptr_ = (void*)(uintptr_t(ptr_) | kDisplaceBit);
+      if (direction == 1) {
+        ptr_ = (void*)(uintptr_t(ptr_) | kDisplaceDirectionBit);
+      }
+    }
+
+    void ClearDisplaced() {
+      ptr_ = (void*)(uintptr_t(ptr_) & ~kDisplaceBit);
+      ptr_ = (void*)(uintptr_t(ptr_) & ~kDisplaceDirectionBit);
+    }
+
+    int GetDisplacedDirection() const {
+      DCHECK(IsDisplaced());
+      return (uintptr_t(ptr_) & kDisplaceDirectionBit) == kDisplaceDirectionBit ? 1 : -1;
+    }
+
+    void Reset() {
+      ptr_ = nullptr;
+    }
+
+    void* GetObject() const {
+      if (IsObject())
+        return Raw();
+      DenseLinkKey* lk = (DenseLinkKey*)Raw();
+      return lk->Raw();
+    }
+
+    void SetObject(void* obj) {
+      ptr_ = (void*)((uintptr_t(ptr_) & kTagMask) | (uintptr_t(obj) & ~kTagMask));
+    }
+
+    DensePtr* Next() {
+      if (!IsLink()) {
+        return nullptr;
+      }
+
+      return &((DenseLinkKey*)Raw())->next;
+    }
+
+    const DensePtr* Next() const {
+      if (!IsLink()) {
+        return nullptr;
+      }
+
+      return &((DenseLinkKey*)Raw())->next;
+    }
+
+   private:
+    void* ptr_ = nullptr;  //
+  };
+
+  struct DenseLinkKey : public DensePtr {
+    DensePtr next;  // could be LinkKey* or Object *.
+  };
+
+  static_assert(sizeof(DensePtr) == sizeof(uintptr_t));
+  static_assert(sizeof(DenseLinkKey) == 2 * sizeof(uintptr_t));
+
+  using LinkAllocator = std::pmr::polymorphic_allocator<DenseLinkKey>;
+  using DenseSetList = std::pmr::vector<DensePtr>::iterator;
+  using ConstDenseSetList = std::pmr::vector<DensePtr>::const_iterator;
+
+  bool Equal(const DensePtr* dptr, const void* ptr) const {
+    if (dptr->IsEmpty()) {
+      return false;
+    }
+
+    return Equal(dptr->GetObject(), ptr);
+  }
+
+  std::pmr::memory_resource* mr() {
+    return entries_.get_allocator().resource();
+  }
+
+  uint32_t BucketId(uint64_t hash) const {
+    return hash >> (64 - capacity_log_);
+  }
+
+  uint32_t BucketId(const void* ptr) const {
+    return BucketId(Hash(ptr));
+  }
+
+  // return a DenseSetList (a.k.a iterator) or end if there is an empty chain found
+  DenseSetList FindEmptyAround(uint32_t bid);
+  void Grow();
+
+  // ============ Pseudo Linked List Functions for interacting with Chains ==================
+  class BucketIterator {
```

Comment:
I am not sure why you called it a bucket. Is not it that a bucket (or a bin) is usually a placeholder in an array, i.e. part of a flat data-structure? Maybe I am wrong, but this was my understanding until now. 

this `BucketIterator` meant to iterate over a linked list, so why not call it `LinkIterator` ?

Reply:
When I referred to a bucket I was referring to the group of entries with the same hash i.e a node with a given hash would be put in the corresponding bucket. I renamed it to `ChainIterator` to avoid confusion